gradle/generation/javacc.gradle

```diff
@@ -0,0 +1,102 @@
+// Add a top-level pseudo-task to which we will attach individual regenerate tasks.
+import static groovy.io.FileType.*
+
+configure(rootProject) {
+  configurations {
+    javacc
+  }
+
+  dependencies {
+    javacc "net.java.dev.javacc:javacc:${scriptDepVersions['javacc']}"
+  }
+
+  task javacc() {
+    description "Regenerate sources for corresponding javacc grammar files."
+    group "generation"
+
+    dependsOn ":lucene:queryparser:javaccParserClassic"
+    dependsOn ":lucene:queryparser:javaccParserSurround"
+    dependsOn ":lucene:queryparser:javaccParserFlexible"
+  }
+}
+
+// We always regenerate, no need to declare outputs.
+class JavaCCTask extends DefaultTask {
+  @Input
+  File javaccFile
+
+  JavaCCTask() {
+    dependsOn(project.rootProject.configurations.javacc)
+  }
+
+  @TaskAction
+  def generate() {
+    if (!javaccFile || !javaccFile.exists()) {
+      throw new RuntimeException("JavaCC input file does not exist: ${javaccFile}")
+    }
+    // Remove old files so we can regenerate them
+    def parentDir = javaccFile.parentFile
+    parentDir.eachFileMatch FILES, ~/.*\.java/, { file ->
+      if (file.text.contains("Generated By:JavaCC")) {
+        file.delete()
+      }
+    }
+    logger.lifecycle("Regenerating JavaCC:\n  from: ${javaccFile}\n    to: ${parentDir}")
+
+    project.javaexec {
+      classpath {
+        project.rootProject.configurations.javacc
+      }
+      main = "org.javacc.parser.Main"
+      args += "-OUTPUT_DIRECTORY=${parentDir}"
+      args += [javaccFile]
+    }
+  }
+}
+
+
+configure(project(":lucene:queryparser")) {
+  task javaccParserClassic(type: JavaCCTask) {
+    description "Regenerate classic query parser from java CC.java"
```

Comment:
I think this is missing the specific description?

Reply:
Copy/paste, I'll fix if Dawid didn't beat me to it.